Hours of Operation Widget

EvanHerman opened the pull request:
Creating an hours of operation widget.

jonathanbardo commented on assets/css/admin.css:
```diff
@@ -93,3 +93,39 @@
+/**
+ * Hours of Operation Widget
+ */
+.wphoow-widget .form div {
```
@EvanHerman Let's use the wpcw prefix for consistency

jonathanbardo commented on contact-widgets.php:
```diff
@@ -2,7 +2,7 @@
  * Plugin Name: Contact Widgets
  * Description: Beautifully display social media and contact information on your website with these simple widgets.
- * Version: 1.3.3
+ * Version: 1.3.4
```
@EvanHerman Don't increment the version number until we decide to release the plugin.

EvanHerman replied:
@jbardo I think the version was incremented because we tagged the previous release when we bumped the tested up to version to 4.7.

jonathanbardo replied:
@EvanHerman That's weird, it shouldn't show off as a change then. Okay moving on!

EvanHerman replied:
@jonathanbardo I think the issue was that the version was bumped in the `master` branch, but the `develop` branch versions didn't get bumped (that was the only difference between the branches) - so I just pulled what we had on master and started working off that so everything was in sync moving forward.

jonathanbardo commented on includes/class-base-widget.php:
```diff
@@ -397,9 +397,98 @@ protected function render_form_textarea( array $field ) {
+
+		$field['name'] = str_replace( 'value', strtolower( $day ), $field['name'] );
+
+		$field['disabled'] = ( $hours['not_open'] ) ? true : false;
```
@EvanHerman No need for the parenthesis here.

jonathanbardo commented on includes/class-base-widget.php:
```diff
@@ -397,9 +397,98 @@ protected function render_form_textarea( array $field ) {
+	 */
+	protected function render_day_input( $field, $day, array $hours ) {
+
+		$field['name'] = str_replace( 'value', strtolower( $day ), $field['name'] );
```
@EvanHerman I think you can align the equal sign of those 2 assignation since they deal with fields elements.

jonathanbardo commented on includes/class-base-widget.php:
```diff
@@ -397,9 +397,98 @@ protected function render_form_textarea( array $field ) {
+
+		$disabled_field = $field['disabled'] ? ' disabled="disabled"' : '';
+
+		echo '<select name="' . esc_attr( $field['name'] . '[open]' ) . '"' . $disabled_field . '>';
```
@EvanHerman Seems like there is a lot of `echo` in this block. It might look nicer if we close the php tag and use html directly.

jonathanbardo commented:
@EvanHerman I feel we might be too restrictive. What about we propose a free text alternative on every field (that way if they close for lunch they can manually add the instruction)?

@fjarrett Any thoughts?

fjarrett commented:
@jonathanbardo I see what you mean, but we need to use microformats (as listed in the AC here https://jira.godaddy.com/browse/WPDEV-167) which will be difficult to display semantically without a day/open/close data pattern to work with.

jonathanbardo commented:
@fjarrett I agree, I wouldn't make it default for sure but I want to be as inclusive as we can. The use case we solved won't work for most European countries so I'm thinking we should at least have the possibility of free text. Also, I think we should add an extra textarea at the end for further details (for example: closed during holidays and 1 month in summer etc).

fjarrett commented:
@jonathanbardo Yeah that's a fantastic idea!

EvanHerman commented:
@jonathanbardo How do you propose we add the text field for free text? Should it be visible at all times, and just overide the `open - close` text if present?

Also, why wouldn't this scale for most European countries?

jonathanbardo commented:
@EvanHerman I know a lot of European countries where they close during lunch time or are totally closed in August for example. With the current implementation it would be impossible for them to express that. 

As for the free text I would say we should add a toggle to add more details and hide it by default.

EvanHerman commented:
@jonathanbardo Right, right. I forgot about the lunch breaks in other countries - good catch 👍 

I will get working on the micro formats, free text and additional info fields!

jonathanbardo commented on includes/class-hours.php:
```diff
@@ -150,7 +168,12 @@ protected function get_fields( array $instance, array $fields = [], $ordered = t
+				'sanitizer'     => function( $value ) { return current_user_can( 'unfiltered_html' ) ? $value : wp_kses_post( stripslashes( $value ) ); },
+				'escaper'       => function( $value ) { return nl2br( apply_filters( 'widget_text', $value ) ); },
+				'form_callback' => 'render_form_textarea',
+				'description' => __( 'Enter additional information about your store.', 'contact-widgets' ),
```
@EvanHerman I think `Enter additional information about your business.` might be more universal.

jonathanbardo commented:
@EvanHerman Travis seems to be failing https://travis-ci.org/godaddy/wp-contact-widgets/jobs/183766078. `store_closed` var is never used in admin.js line 188

EvanHerman commented:
@jonathanbardo The acceptance tests should be passing now. I'll still need to write some acceptance tests for the new Hours of Operation widget when all of the features are built out.

EvanHerman commented on includes/class-base-widget.php:
```diff
@@ -397,9 +397,126 @@ protected function render_form_textarea( array $field ) {
+		$hidden_text_field    = ( ! $hours['custom_text_checkbox'] ) ? ' style="display:none;"' : '';
+		$custom_text_field    = '<input type="text" name="' . $field['name'] . '[custom_text]" id="' . $field['name'] . '[custom_text]" class="widefat custom_text_field" value="' . $hours['custom_text'] . '"' . $hidden_text_field . '/>';
+		$closed_checkbox      = '<input name="' . $field['name'] . '[not_open]" id="' . $field['name'] . '[not_open]" class="js_wpcw_closed_checkbox" type="checkbox" value="1" ' . $this->checked( $hours['not_open'], true ) . '><label for="' . $field['name'] . '[not_open]" class="js_wpcw_closed_checkbox"><small>' . esc_html__( 'Closed', 'contact-widgets' ) . '</small></label>';
+
```
@jonathanbardo This section is a bit verbose. Not sure there is a better way to handle it without introducing additional code, which I didn't want to introduce additional bloat.

EvanHerman commented on includes/class-base-widget.php:
```diff
@@ -397,9 +397,126 @@ protected function render_form_textarea( array $field ) {
+		$field['name'] = str_replace( 'value', strtolower( $day ), $field['name'] );
+
+		$disabled_field = $field['disabled'] ? ' disabled="disabled"' : '';
+		$hidden_field   = $hours['custom_text_checkbox'] ? ' style="display:none;"' : '';
```
@jonathanbardo Looking back on this it may be best to utilize classes for hidden. I'm not a fan of inline styles. Thoughts?

EvanHerman commented on includes/class-base-widget.php:
```diff
@@ -397,9 +397,126 @@ protected function render_form_textarea( array $field ) {
+	 */
+	protected function get_half_hour_time_array() {
+
+		$half_hour_steps = range( 0, 47 * 1800, 1800 );
```
@jonathanbardo @fjarrett What do you think about introducing a filter here to allow users to set increments to whole hours, 15 mins, 10 mins etc.? Do you think that would be beneficial?

EvanHerman commented on includes/class-hours.php:
```diff
@@ -0,0 +1,280 @@
+					$hours = $store_hours['not_open'] ? __( 'Closed', 'contact-widgets' ) : $store_hours['open'] . apply_filters( 'wpcw_hours_seperator', ' - ' ) . $store_hours['closed'];
+					$class = $store_hours['not_open'] ? 'closed' : 'open';
+
+					if ( $store_hours['custom_text_checkbox'] ) {
```
It's worth noting here that when a user has a 'custom text' specified the micro formats are not added to the hours of operation item.

EvanHerman commented:
I can't quite figure out why the acceptance tests are failing either.
https://travis-ci.org/godaddy/wp-contact-widgets/builds/184093250

I can see why in the log, but it looks like it should be passing.

jonathanbardo commented:
@EvanHerman Sometimes Selenium is acting weird. When only 1 builds fails I normally try to restart that particular build. I will find some time to look into why it's acting so weird in travis.

EvanHerman commented:
@jonathanbardo Yea, it was a bit sporadic. It would work every so often after I re-ran the tests, but most of the time it would fail. It seemed like something was timing out and causing the tests to fail.

fjarrett commented:
@EvanHerman @jonathanbardo Love the progress here, not a fan of the UI yet.

It seems that each day could have several controls associated with it, so what do you think about being able to expand/collapse each one? It would also allow the ability to have multiple time blocks on a single day (e.g. 9a-12p, 1p-5p).

<img width="324" alt="screen shot 2016-12-19 at 9 52 29 am" src="https://cloud.githubusercontent.com/assets/522158/21319205/e7b60dc0-c5d1-11e6-89ae-9948ec9337b6.png">

EvanHerman commented:
@fjarrett I'm liking that UI a bit better. I agree, it is a little cluttered. I can run with that and see what I can spin up.

jonathanbardo commented:
@fjarrett That UI does look pretty cool. Adding multiple time block would definitely give people more flexibility.

fjarrett commented:
@jonathanbardo Would multiple time blocks eliminate the need for the "Custom Text" option on each day?

jonathanbardo commented:
@fjarrett Definitely. I would still keep the possibility to add free text at the end though.

EvanHerman commented:
@fjarrett I think this is ready for another review with the new UI in place.

jonathanbardo commented on includes/class-hours.php:
```diff
@@ -196,9 +202,23 @@ protected function get_fields( array $instance, array $fields = [], $ordered = t
+				'label'       => __( 'Current Day on Top', 'contact-widgets' ),
+				'type'        => 'checkbox',
+				'label_after' => true,
+				'show_front_end' => false,
```
@EvanHerman I think it would be nicer to align all values here.

EvanHerman commented on includes/class-base-widget.php:
```diff
@@ -312,6 +312,12 @@ protected function before_form_field( array $field ) {
 	 */
 	protected function render_form_input( array $field ) {
 
+		if ( ! is_admin() ) {
```
@jonathanbardo Whoops, this was left over from some testing I was doing. I was getting the 1 value from the checkbox field printing on the front end of the site, and I was trying to hunt it down.

jonathanbardo replied:
@EvanHerman Glad I found.

fjarrett commented on assets/js/admin.js:
```diff
@@ -166,6 +166,155 @@
+		// Hours of Operation select field toggle
+		$( 'body' ).on( 'change', '.js_wpcw_closed_checkbox', function( e ) {
+
+			var select_fields = $( e.currentTarget ).parents( '.day-container' ).find( 'select' );
```
@EvanHerman As a good rule of thumb, vars that represent element objects should use a `$` prefix, while vars that simple contain data should not. This can be helpful when examining the code later.

fjarrett commented on assets/js/admin.js:
```diff
@@ -166,6 +166,155 @@
+
+			if ( $( this ).is( ':checked' ) ) {
+
+				select_fields.attr( 'disabled', 'disabled' );
```
@EvanHerman The `prop` method should be used instead of `attr`.

fjarrett commented on assets/js/admin.js:
```diff
@@ -166,6 +166,155 @@
+
+			}
+
+			select_fields.removeAttr( 'disabled' );
```
@EvanHerman This should be changed to `.prop( 'disabled', false )`

fjarrett commented on assets/js/wp-hours-widget.js:
```diff
@@ -0,0 +1,62 @@
+
+			if ( ! time_blocks ) {
+
+				$open_sign.addClass( 'closed' ).html( wpcw_hours.closed_string );
```
@EvanHerman I think we can just use the `text()` method on these instead of `html()`.

fjarrett commented on assets/js/wp-hours-widget.js:
```diff
@@ -0,0 +1,62 @@
+
+			}
+
+			return time;
```
@EvanHerman This can be condensed into a one-liner:

```
return ( time < 10 ) ? "0" + time : time;
```

fjarrett commented on assets/js/wp-hours-widget.js:
```diff
@@ -0,0 +1,54 @@
+
+		init: function( e ) {
+
+			var date        = new Date(),
```
@EvanHerman Our date object will need to be instantiated using something like `wpcw_hours.gmt_offset` passed from PHP. Since time blocks are presumably set for the site's timezone, we'll want to use the current time from that timezone when checking for open status.

EvanHerman replied:
wow you're right. Don't know what I was thinking there.

fjarrett commented on assets/js/wp-hours-widget.js:
```diff
@@ -0,0 +1,54 @@
+
+			if ( time_blocks.length ) {
+
+				for ( i = 0; i < time_blocks.length; i++ ) {
```
@EvanHerman I'm about to change the `get_schedule()` method output slightly to use a key/value pair for open/close, so this loop will need to change to:

```js
$.each( time_blocks, function( open, close ) {
```

EvanHerman replied:
@fjarrett 👍 No problem. I can adjust after the change.

fjarrett commented on assets/js/admin.js:
```diff
@@ -161,15 +161,286 @@
+			    $close_select = $parent.find( '.time-block-close' );
+
+			$close_select.children().show();
+			$close_select.find( 'option[value="' + time + '"]' ).prevAll().andSelf().hide();
```
@EvanHerman `.andSelf()` is deprecated and replaced by `.addBack()`

cf. https://api.jquery.com/andself/

fjarrett commented on includes/class-base-widget.php:
```diff
@@ -518,12 +610,33 @@ public function enqueue_scripts() {
+
+		global $_wp_admin_css_colors;
+
+		$active_scheme = $_wp_admin_css_colors[ get_user_option( 'admin_color' ) ];
```
@EvanHerman Bro!!! Nice one. I didn't know this was possible. 💯

fjarrett commented on includes/class-base-widget.php:
```diff
@@ -518,12 +610,33 @@ public function enqueue_scripts() {
 		wp_enqueue_style( 'font-awesome', '//maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css', [], '4.5.0' );
-		wp_enqueue_style( 'wpcw-admin', \Contact_Widgets::$assets_url . "css/admin{$suffix}.css", [ 'font-awesome' ], Plugin::$version );
-		wp_enqueue_script( 'wpcw-admin', \Contact_Widgets::$assets_url . "js/admin{$suffix}.js", [ 'jquery' ], Plugin::$version, true );
+		wp_enqueue_style( 'jquery-timepicker', Plugin::$assets_url . "css/jquery.timepicker{$suffix}.css", [ 'font-awesome' ], '1.11.9' );
```
@EvanHerman Since the timepicker is used exclusively by the hours widget, it shouldn't go in the base widget. Instead, we should use method inheritance inside `class-hours.php`. Additionally, we should move `font-awesome` to `class-social.php`.

Another thing to try would be to add an arg to the `enqueue_scripts` method that would allow localize script data to be passed to from every inherited method back to the parent base.

The code below is untested, but it's an example of what I mean:

```php
/* class-hours.php */

public function enqueue_scripts( $localize_script_data = [] ) {

	$localize_script_data = [
		'time_format' => get_option( 'time_format' ),
	];

	parent::enqueue_scripts( $localize_script_data );

	// Stuff here the hours widget needs to enqueue, like jquery-timepicker

}
```

```php
/* class-base-widget.php */

public function enqueue_scripts( $localize_script_data = [] ) {

	// Stuff here every widget needs to enqueue, like wpcw-admin

	static $data = [];

	$data = array_merge_recursive( $data, $localize_script_data );

	if ( $data ) {

		wp_localize_script( 'wpcw-admin', 'wpcw_admin', $data );

	}

}
```

fjarrett replied:
Well, since every widget shares the same JS file that expects the `wpcw_admin` global to always contain certain values, maybe the localize script data thing isn't necessary after all.

But the general idea of using inheritance for enqueuing other scripts and styles like `jquery-timepicker` and `font-awesome` still applies.

EvanHerman commented:
@fjarrett I believe in the latest commit I incorporated what you were referencing. I also went ahead and bumped font awesome up to 4.7.0, and added the version number as a property so we can easily bump it in future. I know Font awesome 5 is in the works.
